File: Task4/src/Zero.scala
/**
 * Created by Шолохов on 02.02.2015.
 */
case class Zero() extends Expression{
  override def evaluate(values: Map[String, Boolean]): Boolean = false
  override def toString(): String = "0"
  override def equals(o: Any): Boolean = {
    o match {
      case Zero => true
      case _ => false
    }
  }
}

File: Task4/src/Zero.java
public class Zero extends Expression {
    public Zero() {
    }

    public boolean equals(Object other) {
        return other instanceof Zero;
    }

    protected Expression substituteExpression(Expression where, Expression what) {
        return this;
    }

    public String toString() {
        return "0";
    }

    public String toPolishString() {
        return "0";
    }

    public int hashCode() {
        return toPolishString().hashCode();
    }
}
